src/Analyzers/CSharp/Analyzers/UsePatternCombinators/AnalyzedPattern.cs

```diff
@@ -0,0 +1,135 @@
+            {
+                OperatorKind = operatorKind;
+                Value = value;
+            }
+        }
+
+        /// <summary>
+        /// Represents an and/or pattern, constructed from a logical and/or expression.
+        /// </summary>
+        internal sealed class Binary : AnalyzedPattern
+        {
+            public readonly AnalyzedPattern Left;
+            public readonly AnalyzedPattern Right;
+            public readonly bool IsDisjunctive;
+            public readonly SyntaxToken Token;
+
+            internal Binary(AnalyzedPattern leftPattern, AnalyzedPattern rightPattern, bool isDisjunctive, SyntaxToken token, IOperation target) : base(target)
+            {
+                Left = leftPattern;
+                Right = rightPattern;
+                IsDisjunctive = isDisjunctive;
+                Token = token;
+            }
+
+            public static AnalyzedPattern? Create(AnalyzedPattern leftPattern, AnalyzedPattern rightPattern, bool isDisjunctive, SyntaxToken token)
+            {
+                var target = leftPattern.Target;
+                if (!SyntaxFactory.AreEquivalent(target.Syntax, rightPattern.Target.Syntax))
+                    return null;
+
+                return new Binary(leftPattern, rightPattern, isDisjunctive, token, target);
+            }
+        }
+
+        /// <summary>
+        /// Represents a not-pattern, constructed from inequality check or a logical-not expression.
+        /// </summary>
+        internal sealed class Not : AnalyzedPattern
+        {
+            public readonly AnalyzedPattern Pattern;
+
+            private Not(AnalyzedPattern pattern, IOperation target) : base(target)
+                => Pattern = pattern;
+
+            private static BinaryOperatorKind Negate(BinaryOperatorKind kind)
+            {
+                return kind switch
+                {
+                    BinaryOperatorKind.LessThan => BinaryOperatorKind.GreaterThanOrEqual,
+                    BinaryOperatorKind.GreaterThan => BinaryOperatorKind.LessThanOrEqual,
+                    BinaryOperatorKind.LessThanOrEqual => BinaryOperatorKind.GreaterThan,
+                    BinaryOperatorKind.GreaterThanOrEqual => BinaryOperatorKind.LessThan,
+                    var v => throw ExceptionUtilities.UnexpectedValue(v)
+                };
+            }
+
+            public static AnalyzedPattern? Create(AnalyzedPattern? pattern)
+            {
+                return pattern switch
+                {
+                    null => null,
+                    Not p => p.Pattern, // Avoid double negative
+                    Relational p => new Relational(Negate(p.OperatorKind), p.Value, p.Target),
+                    Binary { Left: Not left, Right: Not right } p // Apply demorgans's law
+                        => new Binary(left.Pattern, right.Pattern, !p.IsDisjunctive, p.Token, p.Target),
```

Comment:
i thought we weren't doing demorgans.  or is it just for some very specific cases we are?

Reply:
`!(x != 1 && x != 2)`